### ISSUE TITLE
Leak problem of the DSRunner

### ISSUE THREAD

**Issue by nancheng233** [brainpy/BrainPy#706]
Hello brainpy team:
I found a leak problem in your sample code:
```python
with jax.checking_leaks():
    model = bp.dyn.HH(3, gNa=bp.init.Uniform(min_val=100, max_val=140))
    inputs = np.ones(int(100./ bm.dt)) * 6.  # 100 ms
    runner = bp.DSRunner(model, monitors=['V'])
    runner.run(inputs=inputs)
```

it caused:
```
Exception: Leaked trace DynamicJaxprTrace. Leaked tracer(s):

Traced<ShapedArray(uint32[2])>with<DynamicJaxprTrace>
<DynamicJaxprTracer 2107853421152> is referred to by <RandomState 2106157584272>
<RandomState 2106157584272> is referred to by brainpy._src.math.random.DEFAULT

--------------------
For simplicity, JAX has removed its internal frames from the traceback of the following exception. Set JAX_TRACEBACK_FILTERING=off to include these.
```

I noticed that you also mentioned the memory leak problem of DSRunner, and I use it in my project. My question is that would it make the exit code:-1073741819 (0xC0000005) or taichi error message happen?

For my code:
```python
import jax
import random
import warnings
import brainpy as bp
from brainpy import math as bm
import numpy as np

warnings.filterwarnings("ignore")
random.seed(1)
bm.random.seed(42)
bm.set_dt(1.0)


class Exponential(bp.Projection):
    def __init__(self, pre, post, delay, prob, g_max, tau, E, method, allow_multi_conn):
        super().__init__()
        conn = self.createConn(prob, pre, post, allow_multi_conn)
        self.pron = bp.dyn.FullProjAlignPost(
            pre=pre,
            delay=delay,
            comm=bp.dnn.EventCSRLinear(conn, g_max),
            syn=bp.dyn.Expon(size=post.num, tau=tau, method=method),
            out=bp.dyn.CUBA(),
            post=post
        )

    def createConn_by_prop(self, prob, pre, post, allow_multi_conn):
        conn = bp.conn.FixedProb(prob, pre=pre.num, post=post.num, seed=42, allow_multi_conn=allow_multi_conn)
        return conn

    def createConn(self, prob, pre, post, allow_multi_conn):
        if isinstance(prob, float) and 0 <= prob <= 1:
            conn = self.createConn_by_prop(prob, pre, post, allow_multi_conn)
        else:
            print(f"conn is of unsopported type")
        return conn


class EINet(bp.DynamicalSystem):
    def __init__(self, ne, ni, connect_prob, method, allow_multi_conn):
        super().__init__()
        self.neuron_scale = 0.5
        # tauRef = bm.get_dt()*4.+0.0001 # Make sure tauRef always == 5 timesteps
        tauRef = 0 + 0.0001
        self.E = bp.dyn.LifRef(ne, V_rest=-60., V_th=-10., V_reset=-60., tau=20., tau_ref=tauRef,
                               V_initializer=bp.init.Uniform(-100., -15), method=method)
        self.I = bp.dyn.LifRef(ni, V_rest=-60., V_th=-10., V_reset=-60., tau=20., tau_ref=tauRef,
                               V_initializer=bp.init.Uniform(-100., -15), method=method)

        self.E2E = Exponential(self.E, self.E, delay=0.,
                               prob=connect_prob, g_max=1, tau=5., E=0., method=method,
                               allow_multi_conn=allow_multi_conn)
        self.E2I = Exponential(self.E, self.I, delay=0.,
                               prob=connect_prob, g_max=1, tau=5., E=0., method=method,
                               allow_multi_conn=allow_multi_conn)
        self.I2E = Exponential(self.I, self.E, delay=0.,
                               prob=connect_prob, g_max=6, tau=10., E=-80., method=method,
                               allow_multi_conn=allow_multi_conn)
        self.I2I = Exponential(self.I, self.I, delay=0.,
                               prob=connect_prob, g_max=6, tau=10., E=-80., method=method,
                               allow_multi_conn=allow_multi_conn)
        self.ne = ne
        self.ni = ni

    def update(self, inpS, inpE):
        inpS = inpS.astype(bool)
        self.E2E.pron.refs['pre'].spike.value += inpS[:self.ne]
        self.E2I.pron.refs['pre'].spike.value += inpS[:self.ne]
        self.I2E.pron.refs['pre'].spike.value += inpS[self.ni:]
        self.I2I.pron.refs['pre'].spike.value += inpS[self.ni:]

        self.E2E()
        self.E2I()
        self.I2E()
        self.I2I()
        self.E(inpE)
        self.I(inpE)
        return self.E.spike, self.I.spike

    def dump(self, inpS, inpE, nStep, jit=True, save=True):
        runner = bp.DSRunner(self, monitors=['E.spike', 'I.spike', 'E.V', 'I.V', 'E2E.pron.syn.g', 'E2I.pron.syn.g',
                                             'I2E.pron.syn.g', 'I2I.pron.syn.g'], jit=jit)
        inputs = [inpS, bm.ones(nStep) * inpE]
        _ = runner.run(inputs=inputs)


def test(gen_net = True):
    if gen_net is True:

        scope = 144
        nNeuron = scope*1024
        nExc = int(nNeuron/2)
        nInh = int(nNeuron/2)
        nNeuron = nExc+nInh
        connect_prob = 5 / nNeuron
        net = EINet(nExc, nInh,connect_prob,method = "euler", allow_multi_conn= True)

        nStep = 100
        inpE = 100.
        inpS = np.zeros((nStep, nNeuron))
        inpS = inpS.astype(bool)
        # with jax.checking_leaks():
        # with jax.check_tracer_leaks:
        net.dump(inpS,inpE,nStep,save=True,jit=True)


if __name__ == "__main__":
    test()
```

When i used the env:
● python == 3.11
● brainpy == 2.6.0.post20240618[cpu]
● jax and jaxlib == 0.4.31[cpu]

It always reported exit code:-1073741819 (0xC0000005). But when i use the env:
● python == 3.11
● brainpy == 2.6.0.post20241205[cpu]
● jax and jaxlib == 0.4.36[cpu]

There are three kind of situations:
● First, the process runs normally
● Second, just report exit code:-1073741819 (0xC0000005)
● Third, taichi reports:
```
[W 12/09/24 14:46:06.384 52476] [taichi_core_impl.cpp:ti_set_last_error@233] C-API error: (argument out of range) args[1].type
[W 12/09/24 14:46:06.385 95120] [taichi_core_impl.cpp:ti_set_last_error@233] C-API error: (invalid state) Access violation - no RTTI data!
```
By the way, my friend can run this process normally. I used their env but it did not work. Is it my computer broken?

**Comment by chaoming0625** [comment]
Thanks for report. Can you try [brainstate](https://github.com/chaobrain/brainstate/blob/main/examples/106_COBA_HH_2007.py)?

**Comment by chaoming0625** [comment]
A easy way to fix the leaky problem about ``random.DEFAULT`` is reseting the random seed.

For example:

```python

with jax.checking_leaks():
   bm.random.seed()

    model = bp.dyn.HH(3, gNa=bp.init.Uniform(min_val=100, max_val=140))
    inputs = np.ones(int(100./ bm.dt)) * 6.  # 100 ms
    runner = bp.DSRunner(model, monitors=['V'])
    runner.run(inputs=inputs)
```

**Comment by nancheng233** [comment]
I used the latest versions of brainy unit and brainstate to test 5 times in different environments, and they all succeeded. As for random seed, it did not work.

**Comment by chaoming0625** [comment]
Thanks for your feedback.
